File: Backend/app/models/notificacion_model.py
from sqlalchemy import Column, Integer, String, Text, TIMESTAMP, ForeignKey, func
from sqlalchemy.orm import relationship
from app.core.db import Base


class Notificacion(Base):
    __tablename__ = "notificaciones"

    id = Column(Integer, primary_key=True, index=True)
    tipo_destinatario = Column(String(15), nullable=False)
    # usuario | taller | tecnico

    id_destinatario = Column(Integer, nullable=False)
    incidente_id = Column(Integer, ForeignKey("incidentes.id"), nullable=True)

    titulo = Column(String(200), nullable=True)
    cuerpo = Column(Text, nullable=True)
    datos_extra = Column(Text, nullable=True)   # JSON como texto

    enviado_en = Column(TIMESTAMP(timezone=True), nullable=True)
    leido_en = Column(TIMESTAMP(timezone=True), nullable=True)

    estado = Column(String(15), default="pendiente", nullable=False)
    # pendiente | enviado | fallido | leido

    creado_en = Column(TIMESTAMP(timezone=True), server_default=func.now())

    incidente = relationship("Incidente", back_populates="notificaciones")

File: mobile/lib/features/notificaciones/data/models/notificacion_model.dart
import '../../domain/entities/notificacion_entity.dart';

class NotificacionModel extends NotificacionEntity {
  const NotificacionModel({
    required super.id,
    required super.tipoDestinatario,
    super.incidenteId,
    super.titulo,
    super.cuerpo,
    super.datosExtra,
    required super.estado,
    super.enviadoEn,
    super.leidoEn,
    required super.creadoEn,
  });

  factory NotificacionModel.fromJson(Map<String, dynamic> json) =>
      NotificacionModel(
        id: json['id'],
        tipoDestinatario: json['tipo_destinatario'] ?? 'usuario',
        incidenteId: json['incidente_id'],
        titulo: json['titulo'],
        cuerpo: json['cuerpo'],
        datosExtra: json['datos_extra'] is Map
            ? Map<String, dynamic>.from(json['datos_extra'])
            : null,
        estado: json['estado'] ?? 'pendiente',
        enviadoEn: json['enviado_en'] != null
            ? DateTime.tryParse(json['enviado_en'])
            : null,
        leidoEn: json['leido_en'] != null
            ? DateTime.tryParse(json['leido_en'])
            : null,
        creadoEn: DateTime.parse(json['creado_en']),
      );
}